File: 658. Find K Closest Elements.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @param {number} x
 * @return {number[]}
 */
var findClosestElements = function(arr, k, x) {
    if ( arr == null || arr.length < 2 ) {
        return arr;
    }

    var start = 0;
    var end = arr.length - 1;
    var mid;

    while ( start + 1 < end ) {
        mid = Math.round(( start + end )/2);
        if ( arr[mid] < x ) {
            start = mid;
        } else {
            end = mid;
        }
    }

    var total = 0;
    var result = [];
    while ( start >= 0 && end < arr.length && total < k ) {
        if ( Math.abs(arr[start] - x) <= Math.abs(arr[end] - x) ) {
            result.push(arr[start--]);
        } else {
            result.push(arr[end++]);
        }
        total++;
    }

    while ( start >=0 && total < k ) {
        result.push(arr[start--]);
        total ++;
    }

    while ( end < arr.length && total < k ) {
        result.push(arr[end++]);
        total ++;
    }

    return result.sort(function(a, b){
        return a-b;
    });
};

File: 658. Find K Closest Elements.py
class Solution(object):
    def findClosestElements(self, arr, k, x):
        result = []
        if not arr or len(arr) == 0:
            return result

        l = 0
        h = len(arr) - 1
        while l + 1 < h:
            m = (l + h) / 2
            if arr[m] <= x:
                l = m
            else:
                h = m

        while len(result) < k:
            numL = abs(arr[l] - x) if l >= 0 else None
            numH = abs(arr[h] - x) if h < len(arr) else None

            if numL is not None and numH is not None:
                if numL <= numH:
                    result.append(arr[l])
                    l -= 1
                else:
                    result.append(arr[h])
                    h += 1
            elif numL is not None:
                result.append(arr[l])
                l -= 1
            elif numH is not None:
                result.append(arr[h])
                h += 1
            else:
                break

        return sorted(result)
